src/parser/ir/ir_parser.cc

```diff
@@ -0,0 +1,142 @@
+#include "src/ir/operator.h"
+#include "src/ir/types/type_factory.h"
+#include "src/ir/value.h"
+#include "src/parser/ir/ir_parser_generator_grammar.inc/ir_parser_generator/IrLexer.h"
+#include "src/parser/ir/ir_parser_generator_grammar.inc/ir_parser_generator/IrParser.h"
+
+namespace raksha::ir {
+
+using ir_parser_generator::IrLexer;
+using ir_parser_generator::IrParser;
+
+const Operation& IrProgramParser::ConstructOperation(
+    IrParser::OperationContext& operation_context) {
+  // Operator
+  if (context_.GetOperator(operation_context.ID()->getText()) == nullptr) {
+    context_.RegisterOperator(
+        std::make_unique<Operator>(operation_context.ID()->getText()));
+  }
+  // Grammar rule for attributes is either
+  //   -> ID ':' NUMLITERAL (a number type) or
+  //   -> ID ':' '"'ID'"' (a string type)
+  NamedAttributeMap attributes;
+  if (operation_context.attributeList() != nullptr) {
+    for (IrParser::AttributeContext* attribute_context :
+         operation_context.attributeList()->attribute()) {
+      if (auto* string_attribute_context =
+              dynamic_cast<IrParser::StringAttributeContext*>(
+                  attribute_context)) {
+        attributes.insert(
+            {CHECK_NOTNULL(string_attribute_context)->ID(0)->getText(),
+             Attribute::Create<StringAttribute>(
+                 CHECK_NOTNULL(string_attribute_context)->ID(1)->getText())});
+        continue;
+      }
+      auto* num_attribute_context =
+          dynamic_cast<IrParser::NumAttributeContext*>(attribute_context);
+      int64_t parsed_int = 0;
+      bool conversion_succeeds = absl::SimpleAtoi(
+          CHECK_NOTNULL(num_attribute_context)->NUMLITERAL()->getText(),
+          &parsed_int);
+      CHECK(conversion_succeeds == true);
+      attributes.insert({CHECK_NOTNULL(num_attribute_context)->ID()->getText(),
+                         Attribute::Create<Int64Attribute>(parsed_int)});
+    }
+  }
+  // Grammar rule for value is one of the 2 options below
+  //->ANY
+  //->RESULT (a stored value or a operation result or a block argument)
+  ValueList inputs;
+  if (operation_context.argumentList() != nullptr) {
+    inputs = utils::MapIter<Value>(
+        operation_context.argumentList()->value(),
+        [this](IrParser::ValueContext* value_context) {
+          // Value context parent node is expected to have 0 ID child nodes for
+          // ANY value types
+          if (auto* any_value_context =
+                  dynamic_cast<IrParser::AnyValueContext*>(value_context)) {
+            return Value(value::Any());
+          }
+          auto* named_value_context =
+              dynamic_cast<IrParser::NamedValueContext*>(value_context);
+          const std::string& value_id =
+              CHECK_NOTNULL(named_value_context)->RESULT()->getText();
+          // Operation Result Value
+          auto find_value_id_operation = operation_map_.find(value_id);
+          if (find_value_id_operation != operation_map_.end()) {
```

Comment:
Since we are not dealing with multiple operations, we will not be using this feature but why not just leave it here? Do you think this might cause some issues?

Reply:
We will silently ignore these inputs (and also fail) when we start supporting operation lists. It is better to encode the expected invariant as a `CHECK`.

Reply:
ok. Done!